File: src/theme.py
"""Visual theme: F1 team colours, a dark Plotly template, and app CSS.

Kept in one place so the dashboard and the exported README assets share one look.
"""
from __future__ import annotations

import plotly.graph_objects as go
import plotly.io as pio

# Dashboard palette (dark, high-contrast).
BG = "#0d1117"
PANEL = "#161b22"
GRID = "#2a323d"
INK = "#e6edf3"
MUTED = "#9aa7b4"
ACCENT = "#ff1801"  # F1 red

# F1 team colours, matched by a substring of the team name (covers 2023–2024 naming).
TEAM_COLORS = {
    "red bull": "#3671C6",
    "ferrari": "#E8002D",
    "mercedes": "#27F4D2",
    "mclaren": "#FF8000",
    "aston martin": "#229971",
    "alpine": "#2293D1",
    "williams": "#64C4FF",
    "rb": "#6692FF",
    "alphatauri": "#6692FF",
    "racing bulls": "#6692FF",
    "sauber": "#52E252",
    "kick": "#52E252",
    "alfa romeo": "#C92D4B",
    "haas": "#B6BABD",
}
_FALLBACK = ["#ff1801", "#00d2ff"]


def team_color(team: str | None, index: int = 0) -> str:
    """Return a hex colour for a team name, falling back to a bright default."""
    if team:
        t = str(team).lower()
        for key, col in TEAM_COLORS.items():
            if key in t:
                return col
    return _FALLBACK[index % len(_FALLBACK)]


def driver_colors(team_a: str | None, team_b: str | None) -> tuple[str, str]:
    """Two distinct colours for a driver pairing.

    Uses each driver's team colour; if they share a team (or colour), the second is nudged so the
    two traces stay distinguishable.
    """
    ca = team_color(team_a, 0)
    cb = team_color(team_b, 1)
    if ca.lower() == cb.lower():
        cb = "#00d2ff" if ca.lower() != "#00d2ff" else "#ffb000"
    return ca, cb


def register_template() -> str:
    """Register and return the name of a dark Plotly template matching the app."""
    tmpl = go.layout.Template()
    tmpl.layout = go.Layout(
        paper_bgcolor="rgba(0,0,0,0)",
        plot_bgcolor="rgba(0,0,0,0)",
        font=dict(color=INK, family="Inter, system-ui, sans-serif", size=13),
        title=dict(font=dict(color=INK, size=16)),
        xaxis=dict(gridcolor=GRID, zerolinecolor=GRID, linecolor=GRID),
        yaxis=dict(gridcolor=GRID, zerolinecolor=GRID, linecolor=GRID),
        legend=dict(bgcolor="rgba(0,0,0,0)"),
        margin=dict(l=56, r=20, t=48, b=44),
    )
    pio.templates["f1dark"] = tmpl
    return "f1dark"


# App CSS: web fonts, hidden Streamlit chrome, hero header, cards, driver chips.
CSS = """
<style>
@import url('https://fonts.googleapis.com/css2?family=Rajdhani:wght@500;600;700&family=Inter:wght@400;500;600&family=JetBrains+Mono:wght@500&display=swap');

:root { --accent:#ff1801; --panel:#161b22; --ink:#e6edf3; --muted:#9aa7b4; --grid:#2a323d; }

.stApp { background:
    radial-gradient(1200px 600px at 100% -10%, rgba(255,24,1,0.08), transparent 60%),
    #0d1117; }
header[data-testid="stHeader"] { background: transparent; }
#MainMenu, footer, [data-testid="stToolbar"] { visibility: hidden; }

html, body, [class*="css"] { font-family: 'Inter', system-ui, sans-serif; }
h1, h2, h3, .mt-display { font-family: 'Rajdhani', sans-serif; letter-spacing: .5px; }

/* Hero */
.mt-hero { padding: 6px 0 2px; border-bottom: 1px solid var(--grid); margin-bottom: 6px; }
.mt-hero h1 { font-size: 2.15rem; font-weight: 700; margin: 0; line-height: 1.05;
    text-transform: uppercase; }
.mt-hero .bar { height: 4px; width: 96px; margin-top: 8px; border-radius: 3px;
    background: linear-gradient(90deg, var(--accent), #ff6a3d); }
.mt-hero p { color: var(--muted); margin: 8px 0 0; font-size: .92rem; }

/* Driver matchup chips */
.mt-matchup { display:flex; align-items:center; gap:14px; margin: 4px 0 22px; flex-wrap: wrap; }
.mt-chip { display:flex; align-items:center; gap:11px; background: var(--panel);
    border:1px solid var(--grid); border-radius:12px; padding:10px 18px; min-width:150px; }
.mt-chip .dot { width:12px; height:12px; border-radius:50%; flex:0 0 auto; }
.mt-chip-txt { display:flex; flex-direction:column; gap:2px; line-height:1.1; }
.mt-chip .abbr { font-family:'Rajdhani',sans-serif; font-weight:700; font-size:1.25rem;
    letter-spacing:1px; }
.mt-chip .team { color: var(--muted); font-size:.78rem; }
.mt-vs { font-family:'Rajdhani',sans-serif; font-weight:700; color: var(--muted);
    font-size:1.05rem; padding:0 2px; }

/* Metric cards */
[data-testid="stMetric"] { background: var(--panel); border:1px solid var(--grid);
    border-radius:12px; padding:12px 16px; }
[data-testid="stMetricValue"] { font-family:'JetBrains Mono', monospace; }

/* Tabs */
.stTabs [data-baseweb="tab-list"] { gap: 4px; }
.stTabs [data-baseweb="tab"] { font-family:'Rajdhani',sans-serif; font-weight:600;
    letter-spacing:.5px; text-transform:uppercase; }

/* Sidebar */
[data-testid="stSidebar"] { background: #0b0f14; border-right:1px solid var(--grid); }

.block-container { padding-top: 2.2rem; max-width: 1200px; }
</style>
"""


def hero(title: str, subtitle: str) -> str:
    """Return HTML for the hero header block."""
    return (
        f'<div class="mt-hero"><h1>{title}</h1><div class="bar"></div>'
        f'<p>{subtitle}</p></div>'
    )


def matchup(abbr_a: str, team_a: str | None, color_a: str,
            abbr_b: str, team_b: str | None, color_b: str) -> str:
    """Return HTML for the two-driver matchup chips."""
    def chip(abbr, team, color):
        return (f'<div class="mt-chip"><span class="dot" style="background:{color}"></span>'
                f'<span class="mt-chip-txt"><span class="abbr">{abbr}</span>'
                f'<span class="team">{team or ""}</span></span></div>')
    return (f'<div class="mt-matchup">{chip(abbr_a, team_a, color_a)}'
            f'<span class="mt-vs">VS</span>{chip(abbr_b, team_b, color_b)}</div>')

File: web/src/theme.ts
/** Shared visual tokens and F1 team colours (mirrors src/theme.py). */
import type { Config, Layout } from 'plotly.js';

export const BG = '#0a0b0d';
export const PANEL = '#111317';
export const PANEL2 = '#171a1f';
export const GRID = '#23272e';
export const INK = '#e8eaed';
export const MUTED = '#98a0a8';
export const FAINT = '#6a7178';
export const ACCENT = '#e10600';

/** Tyre compound colors (single source; used by stint bars and the degradation chart). */
export const COMPOUND_COLORS: Record<string, string> = {
  SOFT: '#ff2d55',
  MEDIUM: '#ffd60a',
  HARD: '#e8eaed',
  INTERMEDIATE: '#30d158',
  WET: '#0a84ff',
};

export function compoundColor(compound: string): string {
  return COMPOUND_COLORS[compound.toUpperCase()] ?? MUTED;
}

/** Text color that stays readable on a compound-colored fill. */
export function compoundText(compound: string): string {
  return ['MEDIUM', 'HARD'].includes(compound.toUpperCase()) ? BG : INK;
}

/** '#rrggbb' + alpha -> 'rgba(...)'. */
export function hexAlpha(hex: string, alpha: number): string {
  const n = parseInt(hex.slice(1), 16);
  return `rgba(${(n >> 16) & 255}, ${(n >> 8) & 255}, ${n & 255}, ${alpha})`;
}

const TEAM_COLORS: Record<string, string> = {
  'red bull': '#3671C6',
  ferrari: '#E8002D',
  mercedes: '#27F4D2',
  mclaren: '#FF8000',
  'aston martin': '#229971',
  alpine: '#2293D1',
  williams: '#64C4FF',
  rb: '#6692FF',
  alphatauri: '#6692FF',
  'racing bulls': '#6692FF',
  sauber: '#52E252',
  kick: '#52E252',
  'alfa romeo': '#C92D4B',
  haas: '#B6BABD',
};
const FALLBACK = ['#ff1801', '#00d2ff'];

export function teamColor(team: string | null | undefined, index = 0): string {
  if (team) {
    const t = team.toLowerCase();
    for (const [key, col] of Object.entries(TEAM_COLORS)) {
      if (t.includes(key)) return col;
    }
  }
  return FALLBACK[index % FALLBACK.length];
}

/** Two distinct colours for a driver pairing (nudged apart if same team). */
export function driverColors(teamA?: string | null, teamB?: string | null): [string, string] {
  const a = teamColor(teamA, 0);
  let b = teamColor(teamB, 1);
  if (a.toLowerCase() === b.toLowerCase()) b = a.toLowerCase() !== '#00d2ff' ? '#00d2ff' : '#ffb000';
  return [a, b];
}

/** Base Plotly layout: recessive grid, left-aligned muted titles, unified crosshair hover. */
export function baseLayout(title: string, height = 340): Partial<Layout> {
  const tickfont = { family: "'JetBrains Mono', ui-monospace, monospace", size: 11, color: FAINT };
  return {
    title: { text: title.toUpperCase(), font: { color: MUTED, size: 12 }, x: 0, xanchor: 'left' },
    paper_bgcolor: 'rgba(0,0,0,0)',
    plot_bgcolor: 'rgba(0,0,0,0)',
    font: { color: INK, family: 'Inter, system-ui, sans-serif', size: 13 },
    xaxis: { gridcolor: GRID, zeroline: false, linecolor: 'rgba(0,0,0,0)', tickfont,
             showspikes: true, spikecolor: FAINT, spikethickness: 1, spikedash: 'dot', spikemode: 'across' },
    yaxis: { gridcolor: GRID, zeroline: false, linecolor: 'rgba(0,0,0,0)', tickfont },
    legend: { bgcolor: 'rgba(0,0,0,0)', font: { size: 10, color: FAINT } },
    hovermode: 'x unified',
    hoverlabel: { bgcolor: PANEL2, bordercolor: '#2e333b', font: { color: INK, family: 'Inter, system-ui, sans-serif', size: 12 } },
    margin: { l: 56, r: 52, t: 40, b: 44 },
    height,
  };
}

export const PLOT_CONFIG: Partial<Config> = { displayModeBar: false, responsive: true };
